solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -118,10 +122,22 @@
   private final int queryResultMaxDocsCached;
   private final boolean useFilterForSortedQuery;
 
+  /** Special-case cache to handle the lazy-init of {@link #liveDocs}. */
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  private final CompletableFuture<BitDocSet>[] liveDocsCache = new CompletableFuture[1];
```

Comment:
I would declare it to be simply Future and not specifically CompletableFuture as it's an implementation detail that can be local to the only method that uses it.

BTW is any of that SuppressWarnings actually appropriate here?  I don't see either occuring.  There was a big effort to scrub the codebase of needless suppressions and also declaring things better so that we needn't have a warning in the first place.

Reply:
Both SuppressWarnings are necessary with the "internal-one-element-array-for-synchromization" approach, because of the inability to do parameterized array creation. Is there another alternative that I'm missing there?

I'd be ok with Future if you prefer; the only benefit I can see to CompletableFuture on the instance variable (which is private, so arguably is itself an implementation detail, albeit in a very large class :slightly_smiling_face:) is that it provides the `join()` method to get the computation results without forcing us to explicitly handle checked exceptions from the `get()` method.

Reply:
>Both SuppressWarnings are necessary

to be more specific: gradle won't compile the project unless I add these two SuppressWarnings.

Reply:
You can do with only 1 suppression if you use `Array.newInstance` - there's an example in ZkController

Reply:
This thread should also be addressed by b3b38a1b2e280febd6a4a86464fad36bea4c56b8